### DISCUSSION TITLE
How to initialize "global" variables before @InitLocalVariables ?

### DISCUSSION THREAD

**Discussion by mora1n** [thelfer/tfel#554] [Q&A] [answered]
Hello everyone, I would like to define a variable which has its default value and can be changed in @InitLocalVariables. Is there any way to realize it ?

**Comment by mora1n** [comment]
For now I realize it as:
1. Define 2 local variables "flag" and "val": 
> @ LocalVariable real flag;
> @ LocalVariable real val;

2. Normally, flag = 0. but sometimes flag > 0. for example: flag = 1.e-300
3. In @ InitLocalVariables,
> if (flag < 1.e12) then (set default value to "val" and set flag = 1.e13), else we change the value of "val".

Is there any better way ?

**Reply by thelfer** [reply]
Hi @Mora1n,

By default, local variables are **not** initialized and may have any randoma value dependening on what was in memory, this is precisely the purpose of ``@InitLocalVariables``.

Local variables are not persistent and are reset **every time** the behaviour is called (i.e. at each iterations of the mechanical equilibrium).

Local variables do not seem suitable for your use case and maybe auxiliary state variable is a better choice, but I can't be sure at this stage (parameters could be a good choice also for instance)  because I need more information about what you want to do.

Could you precise exactly what is the usage of `flag` and `val` ? In particular, are those variables global or associated to each Gauss point ?

**Reply by mora1n** [reply]
Hello thelfer,
I'd like to use `flag` to record the loading status (i.e. loading or unloading) of each Gaussian point. At first the `flag` is set to 1 to represent the loading status, and then the generalized shear stress (which is what I chose here) is compared between two time steps to determine whether it is loading or unloading, and once it is unloading the `flag` is set to 0 and not modified, and the different statuses result in different Young's modulus. So I need to define a `flag` before `@InitLocalVariable` and set it to 1. Here the `val` are set to an extremely large value in order to initialize `flag` only once at the first time (normally it is very difficult to get a large randoma value).

**Reply by mora1n** [reply]
Hi @thelfer,
I found that `@Parameter` (`@Parameter real flag=1;`, and `flag` can be changed in `@InitLocalVariable` !) is exactly a solution to my problem. Thank you !

**Reply by thelfer** [reply]
@Mora1n Beware that you in that case, you change the value of a copy of a global variable that you can't change, i.e. **every** time the behaviour is called, `flag` is reset to `1`. I am not sure that this is what you want to do.
To be honest, parameters shall have been declared as constant variables. This was not done for

**Reply by mora1n** [reply]
@thelfer You are right, now I haven't found a solution yet...

**Reply by thelfer** [reply]
@Mora1n I still don't know what you want to do :)

**Reply by thelfer** [reply, marked as answer]
@Mora1n I did not see your anwer, sorry

> I'd like to use flag to record the loading status (i.e. loading or unloading) of each Gaussian point. At first the flag is set to 1 to represent the loading status, and then the generalized shear stress (which is what I chose here) is compared between two time steps to determine whether it is loading or unloading, and once it is unloading the flag is set to 0 and not modified, and the different statuses result in different Young's modulus. So I need to define a flag before @InitLocalVariable and set it to 1. Here the val are set to an extremely large value in order to initialize flag only once at the first time (normally it is very difficult to get a large randoma value).

Since you want a value a each Gauss point, the better is to use an auxiliary state variable. An auxiliary state variable appears from the calling solver as a standard internal state variable. Most solvers allow to initialize such variable (if yours don't, the default value is 0).

From what I understand, you can do something like:

~~~~
@AuxiliaryStateVariable real flag;

@InitLocalVariables{
   // This is called before the behaviour integration

   // Do what you have to do depending on the
   // flag value (which contains either the initial value
  //  at the first time step or the value from the last time step otherwise)
}

// this is called after the behaviour integration, once the increment are known and state variables updated to their values at the end of the time step
@UpdateAuxiliaryStateVariables{
  // here update flag for the **next** time step
  ....
}
~~~~
